store/models/common.go

```diff
@@ -292,59 +289,94 @@ func (w *WebURL) Scan(value interface{}) error {
 	return nil
 }
 
-// Time holds a common field for time.
-type Time struct {
+// AnyTime holds a common field for time, and serializes it as
+// a json number.
+type AnyTime struct {
 	time.Time
+	Valid bool
 }
 
-// Value returns this instance serialized for database storage.
-func (t Time) Value() (driver.Value, error) {
-	return t.Time, nil
+// NewAnyTime creates a new Time.
+func NewAnyTime(t time.Time) AnyTime {
+	return AnyTime{Time: t, Valid: true}
 }
 
-// Scan reads the database value and returns an instance.
-func (t *Time) Scan(value interface{}) error {
-	var s string
-	switch temp := value.(type) {
-	case []uint8:
-		s = string(temp)
-		newTime, err := dateparse.ParseAny(s)
-		*t = Time{Time: newTime}
-		return err
-	case time.Time:
-		*t = Time{Time: temp}
-		return nil
-	default:
-		return fmt.Errorf("Unable to convert %v of %T to Time", value, value)
+// AnyTimeFromNull returns an AnyTime from a null.Time.
+func AnyTimeFromNull(t null.Time) AnyTime {
+	return AnyTime{Time: t.Time, Valid: t.Valid}
+}
+
+// MarshalJSON implements json.Marshaler.
+// It will encode null if this time is null.
+func (t AnyTime) MarshalJSON() ([]byte, error) {
+	if !t.Valid {
+		return []byte("null"), nil
 	}
+	return t.Time.MarshalJSON()
 }
 
 // UnmarshalJSON parses the raw time stored in JSON-encoded
 // data and stores it to the Time field.
-func (t *Time) UnmarshalJSON(b []byte) error {
+func (t *AnyTime) UnmarshalJSON(b []byte) error {
 	var n json.Number
 	if err := json.Unmarshal(b, &n); err != nil {
 		return err
 	}
+
+	if len(n) == 0 {
+		t.Valid = false
+		return nil
+	}
+
 	newTime, err := dateparse.ParseAny(n.String())
 	t.Time = newTime.UTC()
+	t.Valid = true
 	return err
 }
 
-// ISO8601 formats and returns the time in ISO 8601 standard.
-func (t Time) ISO8601() string {
-	return t.UTC().Format("2006-01-02T15:04:05Z07:00")
+// MarshalText returns null if not set, or the time.
+func (t AnyTime) MarshalText() ([]byte, error) {
+	if !t.Valid {
+		return []byte("null"), nil
+	}
+	return t.Time.MarshalText()
+}
+
+// UnmarshalText parses null or a valid time.
+func (t *AnyTime) UnmarshalText(text []byte) error {
+	str := string(text)
+	if str == "" || str == "null" {
```

Comment:
"null" seems a bit unusual, where is that coming from?

Reply:
I should elaborate more on the type `models.AnyTime`. I almost deleted it entirely and went the route of having an intermediary type exclusively for marshaling (ala gencodec: https://github.com/fjl/gencodec). Thought that was too dramatic for this PR and just improved what we had, but limited its use to models populated by external JSON input.

It's main purpose is around the unmarshaling of JSON and little else. After that, `time.Time` or `null.Time` are functionally equivalent and it would be nice to reduce the number of Time types.

Love to discuss it further, just know that I refrained from more refactoring due to the size of this PR and wanting a timely merge for postgres support.

Reply:
Agreed that it would be nice to cut down on our time types. Seems like we could make our models.Time handle lots of different text formats and wrap null.Time, giving us the best of all functionality. Seems possible to cut down to a single Time type.

Reply:
From a database perspective, there are at least two required time types:

1. `time.Time`: Provides the time functionality
2. `null.Time`: Can also be null

If you don't mind every time value being nullable, then sure we could consolidate them all behind `models.AnyTime`. Breaking down `AnyTime`s value prop:

3. `models.AnyTime`: Ability to unmarshal a varity of JSON inputs

I'm of the mindset that we can cut 3) `AnyTime` to be a marshaling tool rather than a fundamental type, much like how go-ethereum uses `hexutil.Big` to deserialize hexadecimal numbers into `big.Int` (yes I know we use hexutil.Big` as an actual type atm). And then once we successfully unmarshal, we lean on the more fundamental types `time.Time` and `null.Time`.

Again, if we are ok with every time being nullable, `models.AnyTime` could do it all, I'm just reluctant to place that on fields like `somemodel.CreatedAt` that are autopopulated by `gorm`.

Reply:
Horrible example since`CreatedAt` should be null before it's persisted. 🤷🏾‍♂️Good talk!

Reply:
> AnyTime to be a marshaling tool rather than a fundamental type, much like how go-ethereum uses hexutil.Big to deserialize hexadecimal numbers into big.Int

That sounds cool actually! I sort of dislike having a type for deserialization then using that everywhere - especially a large AnyTime type which supports so many encodings.

Reply:
We're actually doing something like that already with this `logMarshaling` type:
https://github.com/smartcontractkit/chainlink/blob/master/store/models/eth.go#L54-L59

```golang
type Log struct {
	// Consensus fields:
	// address of the contract that generated the event
	Address common.Address `json:"address" gencodec:"required"`
	// list of topics provided by the contract.
	Topics []common.Hash `json:"topics" gencodec:"required"`
	// supplied by the contract, usually ABI-encoded
	Data []byte `json:"data" gencodec:"required"`
....


type logMarshaling struct {
	Data        hexutil.Bytes
	BlockNumber hexutil.Uint64
	TxIndex     hexutil.Uint
	Index       hexutil.Uint
}
```

Notice how `Data` goes from `hexutil.Bytes` to `[]byte`. And it allows you to mark fields as required. This uses the code gen tool gencodec: https://github.com/fjl/gencodec